django/db/models/fields/related_lookups.py

```diff
@@ -24,15 +24,17 @@ def get_normalized_value(value, lhs):
     from django.db.models import Model
     if isinstance(value, Model):
         value_list = []
-        # A case like Restaurant.objects.filter(place=restaurant_instance),
-        # where place is a OneToOneField and the primary key of Restaurant.
-        if getattr(lhs.output_field, 'primary_key', False):
-            return (value.pk,)
         sources = lhs.output_field.get_path_info()[-1].target_fields
         for source in sources:
             while not isinstance(value, source.model) and source.remote_field:
                 source = source.remote_field.model._meta.get_field(source.remote_field.field_name)
-            value_list.append(getattr(value, source.attname))
+            try:
+                key = getattr(value, source.attname)
+            except AttributeError:
+                # A case like Restaurant.objects.filter(place=restaurant_instance),
+                # where place is a OneToOneField and the primary key of Restaurant.
+                key = value.pk
+            value_list.append(key)
```

Comment:
I couldn't came up with a test case. I suggest to be extra safe and `return (value.pk,)` immediately instead of appending to `value_list` just in case it end up containing values from other iterations (e.g. a model instance that happen to have a field/attribute with the same name as one of the looked up `source.attname`).